### PR TITLE
AE-379: Compare oracle and fallback oracle prices

### PR DESCRIPTION
**Pull request by echoAE**
This agent compares the prices of assets returned by the default oracle (typically from Chainlink aggregators) and the fallback oracle (maintained by AAVE).  The threshold for generating an alert is when the percent error between the two oracle prices for an asset is greater than 2%.  Because the fallback oracle is not maintained well, alerts would be generated for every asset with every block.  Therefore, once an alert is generated for an asset, all subsequent alerts within a 24 hour period are suppressed, but counted.  As soon as 24 hours have elapsed, an alert is created containing a count of the number of alerts that would have been generated.

AE-379: Fixing agent to use oracle contract appropriately.  Removing token blacklist because type of price oracle does not matter.
AE-379: Added time delay to alerts to prevent every block from triggering an alert.  The number of alerts each day is reported for each token.
AE-379: Changing from WebSocketProvider to getDefaultProvider.
AE-379: Adding tests and linting fixes.

### CODE REVIEW

**Review by arbitrary-cr0 — CHANGES_REQUESTED** [review]

File: src/compare-oracle-to-fallback/compare-oracle-to-fallback.js [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+  const { numAlertsInLastDay } = tokenAlert;
+  return Finding.fromObject({
+    name: `Large AAVE Price Oracle / Fallback Oracle Difference for ${reserveToken.symbol}`,
+    description:
+    `Token: ${reserveToken.symbol}, Price: ${tokenPrice}, Fallback Price: ${tokenPriceFallback}, `
+    + `Number of alerts in last 24 hours: ${numAlertsInLastDay.toString()}`,
+    alertId: 'AE-AAVE-FALLBACK-ORACLE-DISPARITY',
+    severity: FindingSeverity.Medium,
+    type: FindingType.Suspicious,
+    everestId: '0xa3d1fd85c0b62fa8bab6b818ffc96b5ec57602b6',
```
[the remark this answers is not in the document]
Comment (echoAE) [reply src/compare-oracle-to-fallback/compare-oracle-to-fallback.js]: Done, thanks.

File: src/compare-oracle-to-fallback/compare-oracle-to-fallback.js [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+const { abi: lendingPoolAddressesProviderAbi } = require('../../interfaces/ILendingPoolAddressesProvider.json');
+
+// set up the an ethers provider
+// do not use ethers.providers.WebSocketProvider in production (there is no support)
+// eslint-disable-next-line new-cap
+const jsonRpcProvider = new ethers.providers.getDefaultProvider(getJsonRpcUrl());
+
+// percent error threshold over which we trigger alerts (2% error compared to the price oracle)
+//   percent error = (absolute(oracle - fallbackOracle) / oracle) * 100
+const AAVE_ORACLE_PERCENT_ERROR_THRESHOLD = 2;
```
[the remark this answers is not in the document]
Comment (echoAE) [reply src/compare-oracle-to-fallback/compare-oracle-to-fallback.js]: Moved, thanks.

File: src/compare-oracle-to-fallback/compare-oracle-to-fallback.js [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+      tokenAddress: reserveToken.tokenAddress,
+      tokenPrice,
+      tokenPriceFallback,
+      percentError,
+      numAlertsInLastDay,
+    },
+  });
+}
+
+async function initializeTokensContractsAlerts() {
```
[the remark this answers is not in the document]
Comment (echoAE) [reply src/compare-oracle-to-fallback/compare-oracle-to-fallback.js:52]: A little painful to set up initially, but super helpful for simplifying the tests.

File: src/compare-oracle-to-fallback/compare-oracle-to-fallback.spec.js [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+      const expectedFinding = Finding.fromObject({
+        name: `Large AAVE Price Oracle / Fallback Oracle Difference for ${token.symbol}`,
+        description:
+          `Token: ${token.symbol}, Price: ${assetPriceOracle.toString()}, `
+          + `Fallback Price: ${assetPriceFallback.toString()}, `
+          + `Number of alerts in last 24 hours: ${numAlertsInLastDay.toString()}`,
+        alertId: 'AE-AAVE-FALLBACK-ORACLE-DISPARITY',
+        severity: FindingSeverity.Medium,
+        type: FindingType.Suspicious,
+        everestId: '0xa3d1fd85c0b62fa8bab6b818ffc96b5ec57602b6',
```
[the remark this answers is not in the document]
Comment (echoAE) [reply src/compare-oracle-to-fallback/compare-oracle-to-fallback.spec.js]: This is taken care of by importing the `createAlert(...)` function from the module (which takes care of importing the `everestId` value from the `agent-config.json` file).

File: interfaces/IPriceOracle.json [hunk 5]
```diff
@@ -3,42 +3,242 @@
...
-          "internalType": "uint256",
-          "name": "price",
-          "type": "uint256"
-        }
-      ],
-      "name": "setAssetPrice",
-      "outputs": [],
-      "stateMutability": "nonpayable",
-      "type": "function"
+		{
```
Comment (arbitrary-kdawg) [inline interfaces/IPriceOracle.json:6]: What's going on with the whitespace in this file?

File: src/compare-oracle-to-fallback/compare-oracle-to-fallback.js [hunk 6]
```diff
@@ -0,0 +1,185 @@
...
+const {
+  Finding, FindingSeverity, FindingType, getJsonRpcUrl,
+} = require('forta-agent');
+
+// load required shared types
+const {
+  LendingPoolAddressesProvider: lendingPoolAddressesProvider,
+  ProtocolDataProvider: protocolDataProviderAddress,
+} = require('../../contract-addresses.json');
+const { abi: protocolDataProviderAbi } = require('../../interfaces/AaveProtocolDataProvider.json');
```
Comment (arbitrary-kdawg) [inline src/compare-oracle-to-fallback/compare-oracle-to-fallback.js:12]: interfaces -> abi

File: src/compare-oracle-to-fallback/compare-oracle-to-fallback.js [hunk 7]
```diff
@@ -0,0 +1,185 @@
...
+const { abi: priceOracleGetterAbi } = require('../../interfaces/IPriceOracleGetter.json');
+const { abi: lendingPoolAddressesProviderAbi } = require('../../interfaces/ILendingPoolAddressesProvider.json');
+const { compareOracleToFallback, aaveEverestId } = require('../../agent-config.json');
+
+// set up the an ethers provider
+// do not use ethers.providers.WebSocketProvider in production (there is no support)
+const jsonRpcProvider = new ethers.providers.JsonRpcProvider(getJsonRpcUrl());
+
+// NOTE: this value is imported from agent-config.json
+// percent error threshold over which we trigger alerts (2% error compared to the price oracle)
```
Comment (arbitrary-kdawg) [inline src/compare-oracle-to-fallback/compare-oracle-to-fallback.js:23]: "2% error" - maybe reword this to highlight that 2% is the current config file default, or don't mention the value at all?

File: src/compare-oracle-to-fallback/compare-oracle-to-fallback.js [hunk 8]
```diff
@@ -0,0 +1,185 @@
...
+function createAlert(reserveToken, tokenPrice, tokenPriceFallback, percentError, tokenAlert) {
+  const { numAlertsInLastDay } = tokenAlert;
+  return Finding.fromObject({
+    name: `Large AAVE Price Oracle / Fallback Oracle Difference for ${reserveToken.symbol}`,
+    description:
+    `Token: ${reserveToken.symbol}, Price: ${tokenPrice}, Fallback Price: ${tokenPriceFallback}, `
+    + `Number of alerts in last 24 hours: ${numAlertsInLastDay.toString()}`,
+    alertId: 'AE-AAVE-FALLBACK-ORACLE-DISPARITY',
+    severity: FindingSeverity.Medium,
+    type: FindingType.Suspicious,
```
Comment (arbitrary-kdawg) [inline src/compare-oracle-to-fallback/compare-oracle-to-fallback.js:39]: Suspicious -> Degraded
Severity -> High?

File: src/compare-oracle-to-fallback/compare-oracle-to-fallback.spec.js [hunk 9]
```diff
@@ -0,0 +1,167 @@
...
+      },
+    ];
+
+    // wrap the Array of Tuples in a Promise (which we will resolve immediately)
+    // the original function returns a Promise that is resolved when all contract interactions have
+    // finished
+    return Promise.resolve(tokenContractFallbackAlertTuples);
+  }
+
+  function calculatePercentError(first, second) {
```
Comment (arbitrary-kdawg) [inline src/compare-oracle-to-fallback/compare-oracle-to-fallback.spec.js:43]: Consider making this a shared function since you do the identical operation inline in the agent code

**Review by arbitrary-kdawg — CHANGES_REQUESTED** [review]
- Update README.md
- Address comments

File: interfaces/IPriceOracle.json (hunk 5 above)
Comment (echoAE) [reply interfaces/IPriceOracle.json]: Removed changes to `IPriceOracle.json` that belong in `AaveOracle.json`.

File: src/compare-oracle-to-fallback/compare-oracle-to-fallback.js (hunk 6 above)
Comment (echoAE) [reply src/compare-oracle-to-fallback/compare-oracle-to-fallback.js]: Replaced all reference to `interfaces` with `abi`.

File: src/compare-oracle-to-fallback/compare-oracle-to-fallback.js (hunk 7 above)
Comment (echoAE) [reply src/compare-oracle-to-fallback/compare-oracle-to-fallback.js]: Removed reference to the value.

File: src/compare-oracle-to-fallback/compare-oracle-to-fallback.js (hunk 8 above)
Comment (echoAE) [reply src/compare-oracle-to-fallback/compare-oracle-to-fallback.js]: Good suggestions.  Replaced.

File: src/compare-oracle-to-fallback/compare-oracle-to-fallback.spec.js (hunk 9 above)
Comment (echoAE) [reply src/compare-oracle-to-fallback/compare-oracle-to-fallback.spec.js]: This function and two others have been moved to `agent-setup.js`.

File: src/compare-oracle-to-fallback/agent-setup.js [hunk 10]
```diff
@@ -0,0 +1,103 @@
...
+function calculatePercentError(first, second) {
+  const delta = first.minus(second).absoluteValue();
+  return delta.div(first).multipliedBy(100);
+}
+
+// helper function to create alerts
+function createAlert(reserveToken, tokenPrice, tokenPriceFallback, percentError, tokenAlert) {
+  const { numAlertsInLastDay } = tokenAlert;
+  return Finding.fromObject({
+    name: `Large AAVE Price Oracle / Fallback Oracle Difference for ${reserveToken.symbol}`,
```
Comment (arbitrary-kdawg) [inline src/compare-oracle-to-fallback/agent-setup.js:31]: Looking at the Forta explorer I suggest we make the alert name more concise.  How about:
"Aave Fallback Oracle Price Difference for <token>"

Also use "Aave" not "AAVE"

File: src/compare-oracle-to-fallback/compare-oracle-to-fallback.spec.js [hunk 11]
```diff
@@ -0,0 +1,168 @@
...
+  ...jest.requireActual('./agent-setup'),
+  initializeTokensContractsAlerts: jest.fn().mockResolvedValue(),
+}));
+
+const { createAlert, calculatePercentError } = require('./agent-setup');
+const { provideHandleBlock } = require('./compare-oracle-to-fallback');
+
+const SECONDS_PER_DAY = 86400;
+
+describe('AAVE oracle versus fallback oracle agent', () => {
```
Comment (arbitrary-kdawg) [inline src/compare-oracle-to-fallback/compare-oracle-to-fallback.spec.js:17]: AAVE -> Aave

File: src/compare-oracle-to-fallback/agent-setup.js (hunk 10 above)
Comment (echoAE) [reply src/compare-oracle-to-fallback/agent-setup.js]: Done and done.

File: src/compare-oracle-to-fallback/compare-oracle-to-fallback.spec.js (hunk 11 above)
Comment (echoAE) [reply src/compare-oracle-to-fallback/compare-oracle-to-fallback.spec.js]: Done.

**Comment by echoAE** [comment]
Comments reviewed and README.md updated.

**Review by arbitrary-kdawg — APPROVED** [review]

**Review by y0wl — COMMENTED** [review]
I wrote these a long time ago but never submitted them

**Review by arbitrary-cr0 — APPROVED** [review]

File: interfaces/IPriceOracle.json [hunk 12]
```diff
@@ -3,42 +3,242 @@
...
+		},
+		{
+			"inputs": [],
+			"name": "getFallbackOracle",
+			"outputs": [
+				{
+					"internalType": "address",
+					"name": "",
+					"type": "address"
+				}
```
[the remark this answers is not in the document]
Comment (echoAE) [reply interfaces/IPriceOracle.json]: Fixed.

File: src/compare-oracle-to-fallback/compare-oracle-to-fallback.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (echoAE) [reply src/compare-oracle-to-fallback/compare-oracle-to-fallback.js]: Moved.

File: src/compare-oracle-to-fallback/compare-oracle-to-fallback.js [hunk 13]
```diff
@@ -0,0 +1,189 @@
...
+
+// percent error threshold over which we trigger alerts (2% error compared to the price oracle)
+//   percent error = (absolute(oracle - fallbackOracle) / oracle) * 100
+const AAVE_ORACLE_PERCENT_ERROR_THRESHOLD = 2;
+
+const SECONDS_PER_DAY = 86400;
+
+// helper function to create alerts
+function createAlert(reserveToken, tokenPrice, tokenPriceFallback, percentError, tokenAlert) {
+  const { numAlertsInLastDay } = tokenAlert;
```
[the remark this answers is not in the document]
Comment (echoAE) [reply src/compare-oracle-to-fallback/compare-oracle-to-fallback.js]: Moved.

File: src/compare-oracle-to-fallback/compare-oracle-to-fallback.js [hunk 14]
```diff
@@ -0,0 +1,189 @@
...
+// set up the an ethers provider
+// do not use ethers.providers.WebSocketProvider in production (there is no support)
+// eslint-disable-next-line new-cap
+const jsonRpcProvider = new ethers.providers.getDefaultProvider(getJsonRpcUrl());
+
+// percent error threshold over which we trigger alerts (2% error compared to the price oracle)
+//   percent error = (absolute(oracle - fallbackOracle) / oracle) * 100
+const AAVE_ORACLE_PERCENT_ERROR_THRESHOLD = 2;
+
+const SECONDS_PER_DAY = 86400;
```
[the remark this answers is not in the document]
Comment (echoAE) [reply src/compare-oracle-to-fallback/compare-oracle-to-fallback.js]: Moved.

File: src/compare-oracle-to-fallback/compare-oracle-to-fallback.js [hunk 15]
```diff
@@ -0,0 +1,189 @@
...
+
+// helper function to create alerts
+function createAlert(reserveToken, tokenPrice, tokenPriceFallback, percentError, tokenAlert) {
+  const { numAlertsInLastDay } = tokenAlert;
+  return Finding.fromObject({
+    name: `Large AAVE Price Oracle / Fallback Oracle Difference for ${reserveToken.symbol}`,
+    description:
+    `Token: ${reserveToken.symbol}, Price: ${tokenPrice}, Fallback Price: ${tokenPriceFallback}, `
+    + `Number of alerts in last 24 hours: ${numAlertsInLastDay.toString()}`,
```
[the remark this answers is not in the document]
Comment (echoAE) [reply src/compare-oracle-to-fallback/compare-oracle-to-fallback.js]: It does not, but I added the spaces just now.
